### PR TITLE
Fix Dynamo silently dropping custom jvp of autograd.Function under forward AD

### LINKED ISSUE #192985: Dynamo silently drops custom `jvp` of `autograd.Function` under `torch.func.jvp`, producing wrong tangents
## 🐛 Describe the bug

When `torch.func.jvp` is called inside a `torch.compile`d function, Dynamo inlines the `forward()` of a `torch.autograd.Function` subclass as plain code and **silently discards its custom `jvp` staticmethod**. The functorch interpreter then differentiates the inlined primitives with their default forward-AD rules, so the compiled function returns a *different* tangent than eager — with no error, no warning, and no graph break, even under `fullgraph=True`.

### Minimal repro

The custom `jvp` deliberately returns zeros (not the true derivative of `sin`), making the drop directly observable:

```python
import torch


class SinWithZeroJvp(torch.autograd.Function):
    @staticmethod
    def forward(x):
        return torch.sin(x)

    @staticmethod
    def setup_context(ctx, inputs, output):
        ctx.save_for_forward(*inputs)
        ctx.save_for_backward(*inputs)

    @staticmethod
    def backward(ctx, grad):
        return torch.zeros_like(grad)

    @staticmethod
    def jvp(ctx, dx):
        return torch.zeros_like(dx)


def f(x, dx):
    return torch.func.jvp(SinWithZeroJvp.apply, (x,), (dx,))[1]


x, dx = torch.randn(4), torch.randn(4)
print(f(x, dx))  # zeros: custom jvp respected
print(
    torch.compile(f, backend="aot_eager", fullgraph=True)(x, dx)
)  # cos(x)*dx: custom jvp dropped

# Backward mode is handled correctly for comparison:
xg = x.clone().requires_grad_()
torch.compile(lambda t: SinWithZeroJvp.apply(t).sum())(xg).backward()
print(xg.grad)  # zeros: custom backward respected
```

Output:

```
tensor([0., 0., 0., 0.])
tensor([ 1.3243, -0.0906,  0.5372, -0.7125])
tensor([0., 0., 0., 0.])
```

Reproduces with `backend="eager"`, `"aot_eager"`, and `"inductor"`, on CPU and CUDA, and with `fullgraph=True`.

### Potential Cause

Dynamo already has an explicit check for this case — the "Unsupported custom jvp" graph break in `AutogradFunctionVariable.call_apply` (`torch/_dynamo/variables/misc.py`) — but it is unreachable from forward-mode AD, because the entire protected path is gated on:

```python
if requires_grad and torch.is_grad_enabled():
```

https://github.com/pytorch/pytorch/blob/0c4461ed5d9578cdf8bf67622c83f2df122e8e4b/torch/_dynamo/variables/misc.py#L1113

Forward-mode differentiation carries dual tensors with `requires_grad=False`, so the gate evaluates falsy, and execution falls through to the plain-inline branch at the bottom of `call_apply`, which drops all autograd customization (`jvp`, `vjp`, custom `backward`).

Backward mode is *not* affected: with grad-requiring inputs, the gate is taken and the custom `backward` is correctly captured via the `AutogradFunctionApply` higher-order op (verified below).

### Expected behavior

Either of:

1. The existing "Unsupported custom jvp" / "Unsupported custom vjp" graph breaks fire whenever a functorch forward-AD transform is active (not only when inputs have `requires_grad=True`), so the eager `autograd.Function` machinery — including the custom `jvp` — is used; or
2. the custom `jvp` is captured and used in the compiled graph.

Silently computing a different derivative than eager should not be possible.

### Versions

PyTorch version: 2.14.0.dev20260807+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux 9.6 (Plow) (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-5)
Clang version: 19.1.7 (Red Hat, Inc. 19.1.7-2.el9)
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.11.15 (main, Mar  3 2026, 14:59:53) [Clang 21.1.4 ] (64-bit runtime)
Python platform: Linux-5.14.0-570.116.1.el9_6.x86_64-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA A100 80GB PCIe
Nvidia driver version: 575.57.08
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.9.11.0
/usr/lib64/libcudnn_adv.so.9.11.0
/usr/lib64/libcudnn_cnn.so.9.11.0
/usr/lib64/libcudnn_engines_precompiled.so.9.11.0
/usr/lib64/libcudnn_engines_runtime_compiled.so.9.11.0
/usr/lib64/libcudnn_graph.so.9.11.0
/usr/lib64/libcudnn_heuristic.so.9.11.0
/usr/lib64/libcudnn_ops.so.9.11.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  64
On-line CPU(s) list:                     0-63
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 6326 CPU @ 2.90GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      16
Socket(s):                               2
Stepping:                                6
BogoMIPS:                                5800.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smxest tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp_epp vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               1.5 MiB (32 instances)
L1i cache:                               1 MiB (32 instances)
L2 cache:                                40 MiB (32 instances)
L3 cache:                                48 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-15,32-47
NUMA node1 CPU(s):                       16-31,48-63
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.9.86
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.14.0.dev20260807+cu126
[pip3] triton==3.8.0+git10f6be36

cc @chauhang @penguinwu @Chillee @samdow @kshitij12345 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98

### PR DESCRIPTION
**Pull request by RyanL2**
Fixes #192985.

`AutogradFunctionVariable.call_apply` gated all autograd customization behind `requires_grad and torch.is_grad_enabled()`. Forward-mode duals carry `requires_grad=False`, so under `torch.func.jvp` or a bare `forward_ad.dual_level()` that gate is falsy and tracing fell through to the plain-inline branch, which traces `forward()` as ordinary Python and discards the custom `jvp`. The functorch interpreter then differentiated the inlined primitives with their default forward-AD rules, so the compiled function returned a different tangent than eager with no error, no warning and no graph break, even under `fullgraph=True`.

The existing "Unsupported custom jvp" graph break was nested inside that same gate, so it was unreachable from forward mode. This hoists the check above the gate and fires it whenever forward AD can actually propagate a tangent, which is two conditions: `forward_ad._current_level >= 0` and `torch._C._is_fwd_grad_enabled()`. The level test covers both entry points, since `torch.func.jvp` enters a dual level via `_jvp_with_argnums` and `dual_level()` is one. The forward-grad-mode test matters because `_set_fwd_grad_enabled(False)` makes `make_dual` return a plain tensor, so no `jvp` rule is consulted however many levels are live, and breaking there would be a pure regression; Dynamo models that context manager in `SetFwdGradEnabledContextManager` and mutates the flag for real while tracing, so it is reachable from traced code.

Scanning the functorch interpreter stack for a `TransformType.Jvp` layer was tried and dropped as redundant. Instrumenting `_current_level` alongside the stack across `jvp`, nested `jvp`, `vmap(jvp)`, `jacfwd`, `grad`, `vmap` and bare `dual_level` shows the level is non-negative in every configuration that has a `Jvp` layer on the stack: `_jvp_with_argnums` enters `dual_level` at `JVP_NESTING == 1` and inner nestings inherit it. The level test alone therefore subsumes the scan, including the `vmap(jvp(f))` case. Matching `TransformType.Jvp` rather than `_are_functorch_transforms_active()` was also considered, and is unnecessary for the same reason: `grad` and `vmap` never enter a dual level, so they never satisfy the forward-AD half of the check. (Under `grad` the pre-existing `requires_grad` half still fires, since #193724 made functorch-wrapped inputs report `requires_grad=True`.)

The break decision reads ambient global state, and nothing invalidates it by default. `FUNCTORCH_STACK_MATCH` is installed only when the functorch stack is already non-empty at compile time, `DUAL_LEVEL` only when Dynamo traces `dual_level()` inside the compiled region, and `GlobalStateGuard` covers neither. So a function compiled with no forward AD live keeps an artifact that inlined `forward()`; passing a dual tensor to it under a `dual_level()` entered outside the region reuses that artifact and reproduces the original silent wrongness, returning `cos(x)*t` instead of the custom rule's zeros. This installs `DUAL_LEVEL` at the decision site, and only when the break was declined, so entering a dual level later forces a recompile. Guarding ambiently in `init_ambient_guards` was the alternative; it was rejected because it would pin the dual level for every compilation, whereas this placement costs nothing for code with no custom `jvp`.

The alternative considered for the fix as a whole was routing `apply()` through the `custom_function_call` HOP whenever functorch transforms are active, mirroring what eager `Function.apply` does in C++, which would capture the custom `jvp` rather than break on it and would close the same hole for custom `vmap` rules. That is the better end state but a substantially larger change; a graph break was chosen here because it restores exact eager semantics through fallback and is straightforward to reason about.

One behavioral note for reviewers: a Function defining both a custom `vjp` and a custom `jvp` with grad-requiring inputs now reports the `jvp` break rather than the `vjp` break, since `jvp` is now checked first. Both are graph breaks with identical consequences.

This also deletes eight `test/dynamo_expected_failures/` entries for `TestAutogradFunctionCPU.test_function_returns_input`. Those entries recorded this very bug: the test asserts that `jvp(fn, (a,), (a_t,))` raises eager's "An input that has been returned as-is" error, and under `PYTORCH_TEST_WITH_DYNAMO=1` the dropped custom `jvp` meant nothing was raised. With the graph break the call falls back to eager and raises, so they pass and their entries have to go. The matching `TestAutogradFunctionCUDA` entries are left in place: no `dynamo_wrapped` CI config runs on a GPU runner, so they are never exercised and were not verified here.

### Test plan

```
TORCHINDUCTOR_WINDOWS_TESTS=1 python test/dynamo/test_autograd_function.py
TORCHINDUCTOR_WINDOWS_TESTS=1 python test/dynamo/test_higher_order_ops.py
TORCHINDUCTOR_WINDOWS_TESTS=1 python test/dynamo/test_ctx_manager.py
PYTORCH_TEST_WITH_DYNAMO=1 python test/functorch/test_eager_transforms.py -k test_function_returns_input
```

After rebasing onto main, these were rerun against a local build predating the rebase with #193724's `torch/_dynamo/variables/torch.py` change applied on top, so failures below are tests main added after that build, not regressions from this PR. `test_autograd_function.py`: 105 tests, 4 skipped; all of `AutogradFunctionFunctorchTests` passes, and the 15 failures are all `AutogradFunctionTests` cases added on main for `setup_context` guarding and static-attribute resolution. `test_higher_order_ops.py`: 226 tests, 8 skipped, 1 expected failure, one error in `test_cond_dunder_dict_read_then_outer_mutation` (added on main), and the local `InvalidCxxCompiler` error in `test_vmap_multiply_scalar`, which reproduces with this change reverted. `test_ctx_manager.py`: 144 tests, 34 skipped, two failures in autocast and inference-mode tests added on main. `test_function_returns_input`: all 24 parametrizations pass; before the fix exactly the 8 listed in `dynamo_expected_failures` fail, each at the same `jvp` assertion.

Eight tests were added to `AutogradFunctionFunctorchTests`: the issue repro across the `eager` and `aot_eager` backends, the `fullgraph=True` raise, the bare `dual_level()` variant, two guards against over-breaking (no forward AD active, forward grad mode disabled inside a dual level), two checks that reverse mode still uses the custom `backward` (via `.backward()` and under `torch.func.grad`), plus a regression test that a graph compiled with no forward AD live recompiles when a dual level is entered later. That last one was confirmed to fail without the `DUAL_LEVEL` guard and pass with it.

Before this change the repro returned `cos(x)*dx` where eager returns zeros, on all of `eager`, `aot_eager` and `fullgraph=True`. After it, all three match eager, and `fullgraph=True` raises `Unsupported` rather than returning a wrong number.

The `dynamo_expected_failures` harness short-circuits on win32 (`dynamo_test_failures.find_test_dir` returns `None`), so the runs that confirm the "Unexpected success" behavior were done with that short-circuit lifted locally to match Linux CI.

### Related

The reverse-mode half of the same report, `torch.func.grad` dropping a custom `backward` (#193279), was fixed separately by #193724, which landed first. As that PR anticipated, it made this PR's original `test_custom_jvp_under_func_grad_does_not_graph_break` fail: under `grad` the custom `jvp` now hits the pre-existing `requires_grad` break, so the `fullgraph=True` frame-count assertion no longer holds. After rebasing, that test is replaced by `test_custom_jvp_under_func_grad_uses_custom_backward`, which asserts the gradient value the original could not, since the custom `backward` now survives.

Authored with Claude Code.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by RyanL2** [comment]
I realized that this change is user facing so can I please get the not user facing tag removed.

**Comment by jansel** [comment]
@claude review these changes

**Comment by RyanL2** [comment]
@jansel rebased and resolved the conflicts, and addressed the Claude review feedback. The focused Dynamo tests pass locally; CI is running again now.

Would appreciate another look when you have a chance.

**Comment by jansel** [comment]
@claude review again, who are the right reviewers to bring in on that discussion topic?  Look at git blame and provide a summary of the decision that needs to be made.

**Comment by RyanL2** [comment]
@zou3519, the correctness fix and regression tests are ready. Do you mind if you can take a look when you have a chance?
